solana/pyth2wormhole/client/src/main.rs

```diff
@@ -171,301 +206,214 @@ fn handle_attest(
+                                    break;
+                                }
+                            }
+                            seqno
+                        })
+                        .ok_or_else(|| -> ErrBoxSend {
+                            format!("No seqno in program logs").into()
+                        })?;
 
-    // TODO(2021-03-09): Extract logic into helper functions
-    while daemon || finished_count < batches.len() {
-        finished_count = 0;
-        for (batch_no, state) in batches.iter_mut() {
-            match state.get_status().clone() {
-                Sending { attempt_no } => {
                     info!(
-                        "Batch {}/{} contents (group {:?}): {:?}",
-                        batch_no,
-                        batch_count,
-                        state.group_name,
-                        state
-                            .symbols
-                            .iter()
-                            .map(|s| s
-                                .name
-                                .clone()
-                                .unwrap_or(format!("unnamed product {:?}", s.product_addr)))
-                            .collect::<Vec<_>>()
+                        "Batch {}/{}, group {:?} OK",
+                        batch_no, batch_count4fut, group_name4job
                     );
-
-                    // Send the transaction
-                    let res = rpc_client
-                        .get_latest_blockhash()
-                        .map_err(|e| -> ErrBox { e.into() })
-                        .and_then(|latest_blockhash| {
-                            let tx_signed = gen_attest_tx(
-                                p2w_addr,
-                                &config,
-                                &payer,
-                                state.symbols,
-                                &Keypair::new(),
-                                latest_blockhash,
-                            )?;
-
-                            rpc_client
-                                .send_transaction(&tx_signed)
-                                .map_err(|e| -> ErrBox { e.into() })
-                        });
-
-                    // Individual batch errors mustn't prevent other batches from being sent.
-                    match res {
-                        Ok(signature) => {
+                    // NOTE(2022-03-09): p2w_autoattest.py relies on parsing this println!{}
+                    println!("Sequence number: {}", seqno);
+                    Result::<(), ErrBoxSend>::Ok(())
+                };
+
+                debug!(
+                    "Batch {}/{}, group {:?}: Scheduling attestation job",
+                    batch_no, batch_count4fut, b4fut.group_name
+                );
+                if daemon {
+                    // park this routine until a resend condition is met
+                    loop {
+                        if let Some(reason) = b4fut.should_resend(&rpc4fut).await {
```

Comment:
In my opinion, async/await in Rust makes async polling very cheap, because every `await` is essentially a preemption callback to the runtime which can park this task and do something else until a poll determines the awaited task is done. If anything, I think this code needs rate-limiting in this and some other places. Unfortunately, I wasn't able to come up with a good way to tie all RPC requests to a mandatory rate-limiting delay. At this point I think a global lock on a single RPC instance might be the way to go. Ideally, with a way to tie a rate-limiting delay to every lock release.